### DISCUSSION TITLE
question: how to test side-effect functions safetly?

### DISCUSSION THREAD

**Discussion by Luis-Henriquez-Perez** [aldrichtr/tangld#16] [Q&A]
As I write tests for tangled. I want to test things like `tangld-init` and `tangld-build` and `tangld-install` without having to worry about actually overwritting files I have on my machine.

Ideally, we want to "coerce" these functions from stateful to functional. My idea for doing this is via advising. We advise the file-writing/link-making functions in tangled in such a way that these functions will return lists of filenames created and linked.

**Comment by aldrichtr** [comment]
I am not very familiar with using advice, but that seems like a reasonable way to test, as it sounds like 'mock' functionality available in other language's testing libraries.

Another option could be to run them in a test environment like a container or virtual machine?

**Reply by Luis-Henriquez-Perez** [reply]
Ok. I'll take the former. I'll try to run tests using advice.
